### DISCUSSION TITLE
Automatic grain deactivation (how/when)

### DISCUSSION THREAD

**Discussion by swegele** [dotnet/orleans#8975] [Q&A] [answered]
I've not yet been able to find any discussion as to when a grain gets deactivated (because it has been dormant for awhile).  It is my impression from all the videos I watched about .NET Orleans that the system automatically deactivates dormant grains.  Is there any documentation around how/when that happens.

For instance, I have a grain that represents a connected phone interview with 2 callers.  Calls to it's grain methods normally happen every couple minutes.  But there _**MIGHT**_ be a long answer in the interview where there is no grain method called for say like 10-15 minutes.  So I'm wondering if I would need to do something (like a heartbeat of some kind) to keep the grain aware that it is still alive/active.

EDIT:  To be clear, I'm just looking for any info/documentation for how the silo/system decides to automatically deactivate a grain.  I don't remember seeing any specifics about that functionality / logic.  Like is it hours or minutes or days?  Does that time value change based on memory scarcity / pressure?

**Comment by MickaelThumerel** [comment, marked as answer]
["deactivateonidle"](https://learn.microsoft.com/en-us/dotnet/api/orleans.grain.deactivateonidle?view=orleans-8.1) => Ask explicitly to deactivate the grain as soon as flag as idle (Also Usefull in dev phase to tester state saving and recovery)

[keepaliveattribute](https://learn.microsoft.com/en-us/dotnet/api/orleans.keepaliveattribute?view=orleans-8.0) => Flag the grain to be keep alive untils the cluster close

[Activation Info](https://learn.microsoft.com/en-us/dotnet/orleans/host/configuration-guide/activation-collection?pivots=orleans-7-0) => All the information about grain activation/desactivation

**Reply by swegele** [reply]
@MickaelThumerel   I totally missed that third item "Activation Info" in the docs.  That exactly answered every question I had!

Salient points from the doc:

> However, activation collection only takes into consideration how long a particular grain activation has been idle. **_Memory usage isn't used as a factor_**.

> This time after which an idle grain activation becomes subject to collection is called Collection Age Limit. The **_default Collection Age Limit is 15 minutes, but it can be changed_** 

Thanks so much for that!!
